### DISCUSSION TITLE
Veritabanı Alternatifleri

### DISCUSSION THREAD

**Discussion by fresholia** [evrimagaci/periodum#43] [Ideas]
**Sorun:**
Projede gördüğüm kadarıyla çok fazla veri dolaşımı yok. Dolayısıyla uzak veritabanına da ihtiyacımız yok, kaldı ki MySQL çağımıza göre oldukça eski bir teknoloji. Projeyi kurmaya çalışırken gördüğümde şaşırdım.

**Çözüm Önerisi:**
Eğer gerçekten veritabanına ihtiyaç varsa PostgreSQL kullanan iki teknolojiden birini kullanabiliriz,

Birincisi [Prisma.io](https://prisma.io/), şemalar oluşturmanıza ve bu şemaları veritabanına aktarmanıza, kütüphanesiyle beraber ise veritabanı sorgusu yazmaya gerek kalmadan işlem yapmanıza olanak sağlıyor. Bu sayede ortak senkron şekilde proje geliştirilebilir.

İkincisi ise [Supabase.io](https://supabase.io/), Supabase'in Prisma'ya göre artısı sizin bir veritabanı hostu ayarlamanıza gerek kalmadan sizin için host oluşturuyor ve ENV ortamına yazabileceğiniz token veriyor, token ile beraber Supabase'ye bağlanıp kütüphaneyle veritabanı işlemlerini halledebiliyor, istatistikleri ise Supabase arayüzünden görebiliyorsunuz.

Üçüncü ve son önerim ise bunların hiçbirine gerek olmadığı ve verileri yerel veritabanında tutarak birçok maliyet ve yükten kurtulmamız olabilir. Yerel veritabanını da Prisma ile kullanabiliriz ve şemalar oluşturabiliriz.

**Not:** _Issues sekmesinde de şemalar yok, yanlış hatırlamıyorsam Repository'nin Settings sayfasından şemalar oluşturabiliyorduk bu yüzden Label ekleyemedim, eklerseniz sevinirim._

**Comment by akifbora** [comment]
peki bu projede relational database kullanmaya gerek var mı? nosql daha efektif olmaz mıydı?

**Comment by birkanozer** [comment]
Bence de RDB'ler yerine MongoDB daha uygun olabilir gibi.

**Comment by mtkuyucu** [comment]
3 tane tablosu olan ve nerdeyse her data si statik olan bir yapı icin herhangi bir db gayet uygun bence. Ayrıca mysql in eski olduguna katılmıyorum ve kullanılmasında bence sorun yok. Eger yapıda bir degisiklik olmayacaksa sistemin db katmanının bu sekilde kalmasında bence sorun yok. Sadece query sonuçları belki cache lenebilir.

**Comment by HKskn** [comment]
@mtkuyucu kesinlikle. Db'nin şu anda bir önemi yok. Cache konusunda da bir ihtiyacı yok, in-memory cache uygulandı harici bir cache önerilebilir ama yine de gerek yok :). Veri tarafında pek bir değişiklik olmayacak belki ufak tefek içerik güncellemeleri. Buraya hiç odaklanmanıza gerek yok arkadaşlar. Uygulamanın en önemli eksiği responsive değil. İlk versiyonda  desktop adaptive olarak geliştirdik. Özellikle topluluğun bu yönde desteğine ihtiyaç var.

**Comment by cagrimertbakirci** [comment]
> **Not:** _Issues sekmesinde de şemalar yok, yanlış hatırlamıyorsam Repository'nin Settings sayfasından şemalar oluşturabiliyorduk bu yüzden Label ekleyemedim, eklerseniz sevinirim._

Bu halloldu. Artık PR ve issue için kurallar/templateler bulunuyor.

> Projede gördüğüm kadarıyla çok fazla veri dolaşımı yok.

Veritabanı konusunda şöyle bir fikrimiz var ama tartışmaya devam edebiliriz:

- **Elementler:** Bunlar çok daha az değişecek bilgilere sahipler. Benim önerim, 14 gün boyunca herkesin interneti altına üstüne getirip elementlere ait girilebilecek bütün bilgileri toplayıp, hayal edilebilecek en detaylı element veritabanını ortaya çıkarabiliriz. Ben birkaç ay içerisinde bu noktaya getirdim (@eyupcanakman'ın da epey katkısı oldu). Ama komünite bir üzerinden geçerse belki atladığım, potansiyel olarak elementlere dair bilinen, sistematik veritabanlarını bulabiliriz. O 14 günlük "scavenge hunt" sona erdikten sonra, elementleri JSON olarak bile tutabiliriz diyoruz. 
- **Bileşikler:** Şimdi bu, apayrı bir dünya. Bizim bulduğumuz veritabanı 262.000 küsür bileşik içeriyor; ama sorun, o bileşiklerle ilgili çok kısıtlı bilgi veriyor olması. Takdir edeceğiniz üzere bu kadar çok bileşiği manuel olarak derlemek mümkün değil (çok sağlam bir komünite eforu ile belki anca). Dolayısıyla muhtemelen temel düzeyde (belki şu anda olduğu kadarıyla) bileşikleri bırakıp, sonrasında "En yaygın 5000 bileşiğe ait çok daha detaylı bilgi ekleme" gibi bir çabaya girişebiliriz. Ayrıca projenin ilerleyen kısımlarında "Bileşik Ekle" gibi bir özellik de planlıyoruz (komünite katkısını epey pekiştirecektir bu). Dolayısıyla bileşikler veritabanı (elementlerin aksine) çok daha değişime açık olması gerekiyor gibime geliyor.

Şimdi, işin teknik tarafını maintainer arkadaşlarım daha iyi değerlendirecektir; o konuda bir yorumum yok. Ancak ne olursa olsun, hem biz koddan az anlayan ölümlüler tarafından hem siz coder'lar tarafından rahatlıkla kullanılabilecek bir yapıda olması şu etapta çok iyi olurdu. Bundan kastım, neredeyse Google Sheets kadar kolay bir şekilde hücrelere ekleme çıkarmalar yapabileceğimiz ama tabii etkili bir şekilde kodu çalıştırabilecek bir sistem gerekiyor (en azından iyice oturana dek).

Burada @cangencler'in [AirTable](https://www.airtable.com/) önerisi oldu; fakat bu detaylı bilgiler ışığında o da belki önerisini gözden geçirir veya hâlen iyi bir opsiyon olacağını düşünüyorsa burada o seçeneği anlatabilir bize.

Hepinize katkılarınızdan ötürü çok teşekkürler.

**Edit:** Unutmuşum, veritabanını da ayrı bir açık kaynak proje yapma fikrimiz de var. Zaten yukarıda anlattıklarımdan anlaşılıyor diye umuyorum ama yine de net olarak burada bahsetmek istedim.

**Reply by cangencler** [reply]
Eğer statik bir site düşünüyorsak (ki içinde bulunduğumuz durumda bu en mantıklı çözüm gibi duruyor) verilerimizi her dağıtımdan önce güncelleyip o şekilde yayına almanın en optimal çözüm olduğunu düşünüyorum.

AirTable önerisi insanların(kodlama bilsin bilmesin) kolayca katkıda bulunabilmeleri ve database'i genişletebilmeleri açısından mantıklı. İlk etapta AirTable'ın kendi formları kullanılarak insanların  database'e kolayca katkıda bulunmaları sağlanabilir. Projenin ileri aşamalarında da bu formu periodum.com icine alarak AirTable API uzerinden kolayca yönetebiliriz. Girilen bilgilerin doğruluk kontrolü ve onay sureci de AirTable'ın view özellikleriyle çok daha kolay bir bicimde kontrol edilebilir.

Genel olarak düşüncem bu yönde. 🙂

**Reply by cagrimertbakirci** [reply]
Benim aklıma epey yattı AirTable, sanıyorum tam yukarıda tarif ettiğim gibi bir araç. Biraz daha diğer fikirleri bekleyelim, eğer başka bir araçtan yana fikrimizi değiştirecek bir öneri çıkmazsa database'i AirTable'a taşıyalım derim.

**Reply by mehmetb** [reply]
AirTable ücretli bir çözüm değil mi? Komünitenin erişimine nasıl açacağız? Ücretsiz paketinde "Up to 5 users" diyor. Bence ücretsiz bir çözüm kullanılmalı. Biraz alternatiflere baktım ama gözüme çarpan olmadı. Belki @cangencler'in bildiği ücretsiz (veya açık kaynağa ücretsiz) bir alternatif vardır?

**Reply by nesimtunc** [reply]
MongoDB ile gitmeye karar verdik PubChem'deki verilerin structure'ini relational olarak mirror etmektense minimal ihtiyacimizi JSON olarak kaydetmemiz yeterli gozuktu. @mehmetb 

Detaylari burada yazdik. https://github.com/evrimagaci/periodum/wiki
Backend & API projesini de baslattik. Katkilarinizi bekliyoruz. https://github.com/evrimagaci/periodum-api

**Reply by mtkuyucu** [reply]
@nesimtunc api reposuna ekleyebilir misin beni contributor olarak.
